security/l1-contracts-dev-flow-vision.md

```diff
@@ -0,0 +1,159 @@
+- We do not have a good way to track the full changelog for a given contracts release.
+
+# Proposed Solution
+
+<!-- A high level overview of the proposed solution.
+When there are multiple alternatives there should be an explanation
+of why one solution was picked over other solutions.
+As a rule of thumb, including code snippets (except for defining an external API)
+is likely too low level. -->
+
+The proposed UX upon for L1 smart contract development is described below.
+This UX will be achieved once the ongoing OPCM Upgrades and superchain-ops projects are completed.
+We start from assuming that `op-contracts/v2.0.0` with OPCM Upgrades is shipped as part of Isthmus
+in Cycle 34. We intentionally exclude the upgrade to `op-contracts/v2.0.0` in this section, as it
+may have some unique aspects because it puts this system in place, although we expect it to be
+largely similar to the flow described here.
+
+## Development
+
+A team starts working on a feature. If, and only if, they are 100% sure this feature is going
+into the next release, feature work can happen on contracts directly. Otherwise the feature must be
+developed using the inheritance pattern described in [`smart-contract-feature-development.md`](../smart-contract-feature-development.md).
+
+For example, if you are adding something to the SystemConfig that will ship in the next release, make
+those changes directly to `SystemConfig.sol`. Tests would be added directly to `SystemConfig.t.sol` like normal.
+
+If your feature may not go into the next release, create a child contract such as `contract SystemConfigInterop is SystemConfig`.
+In this contract you can add your feature, and once we know it's ready for release it will be merged
+into the parent contract it inherits from. Tests for `SystemConfigInterop` would live in `SystemConfigInterop.t.sol`
+(TODO flesh out more details on how this should work with OPCM, e.g. do we need an `OPCM*` for each feature, like
+`OPCMInterop`, etc? A simpler approach may be to always use OPCM as the base standard deploy, and write a `interopUpgrade()`
+method that contains the code that will eventually go in `OPCM.upgrade`).
+
+An alternate approach to inheritance is to put code directly into `SystemConfig`, but behind a
+feature flag. This is the approach client code uses so they can ship releases without also shipping
+features that are not production ready. However, in safety-critical systems dead code is typically
+not allowed, so we likely will not allowed this approach for contracts. However, we mention it here
+anyway for completeness.
+
+## Testing
+
+In the above section we described how standard unit tests would be written during development. Another
+aspect is how deploy and upgrade scripts are written and tested during development, as this is a key
+change.
+
+Originally, `Deploy.s.sol` was the script that run as part of test `setUp`, so we can run tests
```

Comment:
Just linking out to this github issue as I don't see it mentioned here: https://github.com/ethereum-optimism/optimism/issues/13331

The same solidity code that is called by `op-deployer` should be called by a test setup script to populate the unit tests. There should be no ffi out to an external process to get JSON to populate the state for solidity unit tests.

Reply:
Yeah +1 on this. The tests should compose the Solidity scripts that OP Deployer uses under the hood. If we want to make defining test chains in Solidity more declarative, we can build that into the Solidity itself.

IMO it's a smell every time we FFI out to something from Solidity.